### ISSUE TITLE
Meteor crashes when HTTP.call receives an error and content is not a string

### ISSUE THREAD

**Issue by lintonye** [meteor/meteor#4847]
## Steps to reproduce
- Create a new project `meteor create http-test`
- Add http package `meteor add http`
- Replace the content of `http-test.js` with the following:

```javascript
if (Meteor.isClient) {

  Template.hello.events({
    'click button': function () {
      Meteor.call('get');
    }
  });
}

if (Meteor.isServer) {
  Meteor.methods({
    'get': function() {
      var url = 'http://www.google.com/1.jpg';
      HTTP.call('GET', url, {npmRequestOptions: {encoding: null}});
    }
  })
}
```
- Start Meteor and click the "Click Me" button in the browser

## Result
- After a few seconds, Meteor crashes with the error below:

```
W20150730-15:02:20.278(-7)? (STDERR) TypeError: Object <!DOCTYPE html>
W20150730-15:02:20.278(-7)? (STDERR) <html lang=en>
W20150730-15:02:20.278(-7)? (STDERR)   <meta charset=utf-8>
W20150730-15:02:20.278(-7)? (STDERR)   <meta name=viewport content="initial-scale=1, minimum-scale=1, width=device-width">
W20150730-15:02:20.278(-7)? (STDERR)   <title>Error 404 (Not Found)!!1</title>
W20150730-15:02:20.278(-7)? (STDERR)   <style>
W20150730-15:02:20.279(-7)? (STDERR)     *{margin:0;padding:0}html,code{font:15px/22px arial,sans-serif}html{background:#fff;color:#222;padding:15px}body{margin:7% auto 0;max-width:390px;min-height:180px;padding:30px 0 15px}* > body{background:url(//www.google.com/images/errors/robot.png) 100% 5px no-repeat;padding-right:205px}p{margin:11px 0 22px;overflow:hidden}ins{color:#777;text-decoration:none}a img{border:0}@media screen and (max-width:772px){body{background:none;margin-top:0;max-width:none;padding-right:0}}#logo{background:url(//www.google.com/images/errors/logo_sm_2.png) no-repeat}@media only screen and (min-resolution:192dpi){#logo{background:url(//www.google.com/images/errors/logo_sm_2_hr.png) no-repeat 0% 0%/100% 100%;-moz-border-image:url(//www.google.com/images/errors/logo_sm_2_hr.png) 0}}@media only screen and (-webkit-min-device-pixel-ratio:2){#logo{background:url(//www.google.com/images/errors/logo_sm_2_hr.png) no-repeat;-webkit-background-size:100% 100%}}#logo{display:inline-block;height:55px;width:150px}
W20150730-15:02:20.279(-7)? (STDERR)   </style>
W20150730-15:02:20.279(-7)? (STDERR)   <a href=//www.google.com/><span id=logo aria-label=Google></span></a>
W20150730-15:02:20.279(-7)? (STDERR)   <p><b>404.</b> <ins>That’s an error.</ins>
W20150730-15:02:20.280(-7)? (STDERR)   <p>The requested URL <code>/1.jpg</code> was not found on this server.  <ins>That’s all we know.</ins>
W20150730-15:02:20.280(-7)? (STDERR)  has no method 'replace'
W20150730-15:02:20.280(-7)? (STDERR)     at makeErrorByStatus (packages/http/httpcall_common.js:10:1)
W20150730-15:02:20.280(-7)? (STDERR)     at Request._callback (packages/http/httpcall_server.js:109:1)
W20150730-15:02:20.280(-7)? (STDERR)     at Request.self.callback (/Users/lintonye/.meteor/packages/http/.1.1.0.qvfsg5++os+web.browser+web.cordova/npm/node_modules/request/request.js:344:22)
W20150730-15:02:20.281(-7)? (STDERR)     at Request.emit (events.js:98:17)
W20150730-15:02:20.281(-7)? (STDERR)     at Request.<anonymous> (/Users/lintonye/.meteor/packages/http/.1.1.0.qvfsg5++os+web.browser+web.cordova/npm/node_modules/request/request.js:1239:14)
W20150730-15:02:20.281(-7)? (STDERR)     at Request.emit (events.js:117:20)
W20150730-15:02:20.282(-7)? (STDERR)     at IncomingMessage.emit (events.js:117:20)
W20150730-15:02:20.281(-7)? (STDERR)     at IncomingMessage.<anonymous> (/Users/lintonye/.meteor/packages/http/.1.1.0.qvfsg5++os+web.browser+web.cordova/npm/node_modules/request/request.js:1187:12)
W20150730-15:02:20.282(-7)? (STDERR)     at _stream_readable.js:944:16
W20150730-15:02:20.282(-7)? (STDERR)     at process._tickCallback (node.js:442:13)
=> Exited with code: 8
=> Your application is crashing. Waiting for file change.
```
- If `{npmRequestOptions: {encoding: null}}` is removed or changed to another encoding such as `base64`, there will be no error.

**Comment by stubailo** [comment]
OK based on that stack trace what happened is:

1. You passed `encoding: null` to the options, which actually worked
2. The URL didn't exist, so Google returned an error code
3. The HTTP package tried to print an error, while printing the error it assumed that the content of the response was a string
4. The content of the response was actually a buffer because of the option you passed

The solution is to have HTTP do something better when the response is not a string.

**Comment by stubailo** [comment]
Here is the code that needs to change: https://github.com/meteor/meteor/blob/eac451c82def092eb57fdf5367e26085e6fca001/packages/http/httpcall_common.js#L10

**Comment by Rebolon** [comment]
https://github.com/meteor/meteor/pull/4940

**Comment by stubailo** [comment]
Closed with commit d6427c4150d9904ab0c48d6a72e4a70b604c378c
